content/2019/Making-things-flow/index.md

```diff
@@ -0,0 +1,79 @@
+---
+title: "Making Things Flow"
+publishDate: 2019-12-01
+draft: true
+description: "Making the behaviour of Vim flow"
+author:
+  github: "george-b"
+  name: "George Brown"
+---
+
+# Making things flow
+
+Changing the behaviour of one's editor is a common subject, but what if the desired behaviour can't be achieved by configuring a setting offered by your editor? Thankfully Vim is an editor which allows powerful scripting by its users.
+
+## Look Ma, no mappings!
+
+There's nothing wrong with mappings. The purpose of this article is to explore more generalised methods whereas mappings generally have a more singular focus. Or to put is another way this article will focus more on the behaviour and response of the editor rather than defining singular imperative actions.
+
+## An example from defaults.vim
+
+Let us begin with an example that can be found in [Vim as-is these days][defaults.vim].
+
+```vim
+" When editing a file, always jump to the last known cursor position.
+" Don't do it when the position is invalid, when inside an event handler
+" (happens when dropping a file on gvim) and for a commit message (it's
+" likely a different one than last time).
+autocmd BufReadPost *
+  \ if line("'\"") >= 1 && line("'\"") <= line("$") && &ft !~# 'commit'
+  \ |   exe "normal! g`\""
+  \ | end
+```
+
+The comment explains the what's being done in the `autocmd` so I won't talk about that. But it's interesting to think about the fact that this is something Vim now ships with. It's been possible and indeed used by others in their configuration before appearing in mainline Vim. I consider this interesting for two reasons, firstly Vim scripting allows us to do this quite easily and secondly the scripting was added to Vim rather than code for an option that would allow users to `set` such behaviour.
+
+## A good primer
+
+Whilst having been a Vim user for some time my first real exposure to the idea of curving Vim's behaviour in a transparent way was this excellent [gist][CCR]. I thoroughly recommend reading through it then coming back here. I actually even made my own [variant][Autoreply].
+
+In any case the effect is the same. Vim's behaviour is altered not because of a setting but because we've scripted a flow. It's something we can use a great deal but we've only ever had to think about it once, when we wrote the script. To contrast we don't have to think "Oh I have a mapping that will show me X and prompt for Y", instead we ask Vim for X as usual it now prompts for Y.
+
+Continuing with a another "call and response" type example having the quickfix window open when it contains results.
+
+Another one of Romain's [gists][instant-grep] showcases this. This is also behaviour is also available in [vim-qf][vim-qf].
```

Comment:
@romainl I've referred to you by you're first name here. Are you OK with this or would you prefer `romainl` or not to be explicitly named at all?

Reply:
Thanks for the notice. Actually, I'd rather not be named at all.